In [1]:
import pandas as pd
from pandas import DataFrame
import numpy as np

In [2]:
df = pd.read_csv("pima-indians-diabetes.csv", header=None)

In [3]:
df.head()

0    1   2   3    4     5      6   7  8
0  6  148  72  35    0  33.6  0.627  50  1
1  1   85  66  29    0  26.6  0.351  31  0
2  8  183  64   0    0  23.3  0.672  32  1
3  1   89  66  23   94  28.1  0.167  21  0
4  0  137  40  35  168  43.1  2.288  33  1

**1) Sobre as quantidades de instâncias e características presentes no dataset, é CORRETO afirmar:**<br>
**2) Considerando apenas a carta utilizando o módulo pandas, quantos tipos de dados diferentes existem no dataset?**

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       768 non-null    int64  
 1   1       768 non-null    int64  
 2   2       768 non-null    int64  
 3   3       768 non-null    int64  
 4   4       768 non-null    int64  
 5   5       768 non-null    float64
 6   6       768 non-null    float64
 7   7       768 non-null    int64  
 8   8       768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


**3) Qual é a idade (coluna 7) média dos pacientes que são diabéticos (coluna 8=1)?**

In [5]:
df[df[8] == 1].describe()

0           1           2           3           4           5  \
count  268.000000  268.000000  268.000000  268.000000  268.000000  268.000000   
mean     4.865672  141.257463   70.824627   22.164179  100.335821   35.142537   
std      3.741239   31.939622   21.491812   17.679711  138.689125    7.262967   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      1.750000  119.000000   66.000000    0.000000    0.000000   30.800000   
50%      4.000000  140.000000   74.000000   27.000000    0.000000   34.250000   
75%      8.000000  167.000000   82.000000   36.000000  167.250000   38.775000   
max     17.000000  199.000000  114.000000   99.000000  846.000000   67.100000   

                6           7      8  
count  268.000000  268.000000  268.0  
mean     0.550500   37.067164    1.0  
std      0.372354   10.968254    0.0  
min      0.088000   21.000000    1.0  
25%      0.262500   28.000000    1.0  
50%      0.449000   36.000000    1.0  
75%      0.728000   44.000000    1.0  
max      2.420000   70.000000    1.0

**4) Dentre os pacientes que NÃO tiveram diabetes (coluna 8 =0), qual é o número máximo de gestações (coluna 0) existentes?**

In [6]:
df[df[8] == 0].describe()

0         1           2           3           4           5  \
count  500.000000  500.0000  500.000000  500.000000  500.000000  500.000000   
mean     3.298000  109.9800   68.184000   19.664000   68.792000   30.304200   
std      3.017185   26.1412   18.063075   14.889947   98.865289    7.689855   
min      0.000000    0.0000    0.000000    0.000000    0.000000    0.000000   
25%      1.000000   93.0000   62.000000    0.000000    0.000000   25.400000   
50%      2.000000  107.0000   70.000000   21.000000   39.000000   30.050000   
75%      5.000000  125.0000   78.000000   31.000000  105.000000   35.300000   
max     13.000000  197.0000  122.000000   60.000000  744.000000   57.300000   

                6           7      8  
count  500.000000  500.000000  500.0  
mean     0.429734   31.190000    0.0  
std      0.299085   11.667655    0.0  
min      0.078000   21.000000    0.0  
25%      0.229750   23.000000    0.0  
50%      0.336000   27.000000    0.0  
75%      0.561750   37.000000    0.0  
max      2.329000   81.000000    0.0

**5) Qual é a afirmação CORRETA sobre o valor 3,0 para a característica número de gestações (coluna 0)?**

In [7]:
df[0].describe()

count    768.000000
mean       3.845052
std        3.369578
min        0.000000
25%        1.000000
50%        3.000000
75%        6.000000
max       17.000000
Name: 0, dtype: float64

In [8]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier

In [9]:
entradas = df.drop(columns=[8], axis = 1)
saida = df[8]

In [10]:
normaliza = MinMaxScaler() #objeto para a normalização
entradas_normalizadas=normaliza.fit_transform(entradas)

**6) Após normalizado, qual é o maior valor presente na coluna concentração de glicose (coluna 1)?**

In [11]:
entradas_normalizadas.max()

1.0

**8) Após dividir as colunas do dataframe entre entrada e saída, aplicar a normalização dos dados como apresentado no enunciado (MinMaxScaler()) e dividir esses dados entre treinamento e teste, aplique o algoritmo KNN. Qual é, aproximadamente, a acurácia do modelo?**

In [12]:
X_train, X_test, y_train, y_test = train_test_split(entradas_normalizadas, saida, test_size=0.30,random_state=42)

**Aplicando KNN**

In [13]:
clf_KNN = KNeighborsClassifier(n_neighbors=5)
clf_KNN.fit(X_train, y_train)

y_pred_KNN = clf_KNN.predict(X_test)

print("Matriz de confusão: \n", confusion_matrix(y_test, y_pred_KNN), "\n")
print("Relatório de classificação: \n", classification_report(y_test, y_pred_KNN), "\n")
print("Acurácia: {}%".format(100*round(accuracy_score(y_test, y_pred_KNN), 2)))

Matriz de confusão: 
 [[118  33]
 [ 41  39]] 

Relatório de classificação: 
               precision    recall  f1-score   support

           0       0.74      0.78      0.76       151
           1       0.54      0.49      0.51        80

    accuracy                           0.68       231
   macro avg       0.64      0.63      0.64       231
weighted avg       0.67      0.68      0.68       231
 

Acurácia: 68.0%


**Após dividir as colunas do dataframe entre entrada e saída, aplicar a normalização dos dados como apresentado no enunciado (MinMaxScaler()) e dividir esses dados entre treinamento e teste, aplique os algoritmos KNN, Árvore de Decisão e MLP. Qual dos algoritmos apresentou maior acurácia?**

**Aplicando Árvore de decisão**

In [14]:
clf_arvore = DecisionTreeClassifier(random_state=1)
clf_arvore.fit(X_train, y_train)

y_pred_arvore = clf_arvore.predict(X_test)

print("Matriz de confusão: \n", confusion_matrix(y_test, y_pred_arvore), "\n")
print("Relatório de classificação: \n", classification_report(y_test, y_pred_arvore), "\n")
print("Acurácia: {}%".format(100*round(accuracy_score(y_test, y_pred_arvore), 2)))

Matriz de confusão: 
 [[106  45]
 [ 25  55]] 

Relatório de classificação: 
               precision    recall  f1-score   support

           0       0.81      0.70      0.75       151
           1       0.55      0.69      0.61        80

    accuracy                           0.70       231
   macro avg       0.68      0.69      0.68       231
weighted avg       0.72      0.70      0.70       231
 

Acurácia: 70.0%


**Aplicando MLP**

In [15]:
clf_mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 10), random_state=1)
clf_mlp.fit(X_train, y_train)

y_pred_mlp = clf_mlp.predict(X_test)

print("Matriz de confusão: \n", confusion_matrix(y_test, y_pred_mlp), "\n")
print("Relatório de classificação: \n", classification_report(y_test, y_pred_mlp), "\n")
print("Acurácia: {}%".format(100*round(accuracy_score(y_test, y_pred_mlp), 2)))

Matriz de confusão: 
 [[124  27]
 [ 32  48]] 

Relatório de classificação: 
               precision    recall  f1-score   support

           0       0.79      0.82      0.81       151
           1       0.64      0.60      0.62        80

    accuracy                           0.74       231
   macro avg       0.72      0.71      0.71       231
weighted avg       0.74      0.74      0.74       231
 

Acurácia: 74.0%


/opt/conda/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:471: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


<hr>
<h1>Respostas</h1>
<br>Pergunta 1 - Existem 768 instâncias e 9 características
<br>Pergunta 2 - 2 tipos diferentes de dados
<br>Pergunta 3 - 37,06 anos
<br>Pergunta 4 - 13
<br>-->> Pergunta 5 - Indica que 50% dos valores presentes nessa coluna são maiores que 3
<br>Pergunta 6 - 1
<br>Pergunta 7 - A criação dos dataframes só ocorre quando é utilizado o comando pandas.read_csv()
<br>Pergunta 8 - 0,68
<br>Pergunta 9 - MLP
<br>Pergunta 10 - Está implementada de maneira equivocada, pois não está acompanhada da tag “end“ <\br>
<br>-->> Pergunta 11 - Substituindo <label for="age">Idade</label> <input type="text" id="age"> por <label>Idade <input type="text"></label> não teríamos prejuízo no funcionamento do código.
<br>-->> Pergunta 12 - Indica qual URL deve chamar a função desejada.
<br>-->> Pergunta 13 - 
<br>-->> Pergunta 14 - Criar um objeto tupla a partir do mapeamento de um iterable
<br>Pergunta 15 - Capturar dados através de um formulário HTML, aplicar previsão a partir de um modelo e retornar a resposta via web
<hr>